### ISSUE TITLE
Parallel dispatches aren't actually reference count neutral

### ISSUE THREAD

**Issue by dhollman** [kokkos/kokkos#2072]
Consider the following code:

```c++
    auto data = Kokkos::View<Kokkos::View<double>>("view of view");

    {
      auto inner = Kokkos::View<double>("inner");

      Kokkos::parallel_for(1,
        KOKKOS_LAMBDA (int i) {
	  inner() = 3.14;
          data() = inner;
        }
      );
    }

    auto new_inner = Kokkos::View<double>("new inner");

    Kokkos::parallel_for(1,
      KOKKOS_LAMBDA (int i) {
	new_inner() = 2.71828;
	printf("3.14 == %g ???\n", data()());
      }
    );
```

On both the `Serial` and `Cuda` backends, this prints `3.14 == 2.71828 ???`, which is because the `new_inner` gets allocated right on top of `inner`.  In a more complicated code, this would probably just segfault.  Is there anything I'm doing here that violates the Kokkos programming model? If so, what?  If not, this should work, or we should add restrictions to the programming model that disallow this.  The `View` assignment operator is device marked, even for managed views, so presumably we intend for assignment on the device to work.  Either way, as currently implemented, this is a bug.

**Comment by mhoemmen** [comment]
Would we expect the following to work correctly?
```c++
View<int*> view0 ("view0", 42);
View<int*> view1 ("view1", 42);
Kokkos::parallel_for(1, KOKKOS_LAMBDA(int i) { view0 = view1; });
```
I would say no -- it's not ok to update some managed View that lives outside the parallel region.  However, the following should be OK:
```c++
View<int*> view0 ("view0", 42);
View<int*> view1 ("view1", 42);
Kokkos::parallel_for(1, KOKKOS_LAMBDA(int i) { 
    // Whether or not reference counting happens inside here,
    // none of the Views or subviews created here may escape
    // the parallel region in which they were created.
    auto view0_sub = Kokkos::subview (view0, std::pair {some_int, some_int + 7});
    auto view1_sub = Kokkos::subview (view1, std::pair {some_other_int, some_other_int + 3});
    view0_sub = view1_sub;
  });
```

**Comment by dhollman** [comment]
@mhoemmen But the first example doesn't compile, because `view0` is captured as `const`. I'd be fine if the example I gave didn't compile, but it does. And runs. And gives the wrong answer. Silently. It's just a landmine waiting for someone to step on it.

**Comment by mhoemmen** [comment]
@dhollman I'm with you there -- it's not cool that the example compiles and runs without warnings, but silently gives the wrong answer.  It also breaks users' model of thread safety ("only one thread could ever touch this entry of the outer View at a time, so why can't I do this?").

**Comment by dhollman** [comment]
I think the problem is that we turn off reference counting globally in a sledgehammer fashion.  We can turn off reference counting in the copy constructor (and tell people that placement `new` is not allowed), but in the assignment operator we still need to increment/decrement counts correctly (even on the GPU).  I haven't thought through all of the corner cases, but I think this should address the problem described here and pretty much any reasonable related problems.

**Comment by dhollman** [comment]
Changed the title to reflect this solution

**Comment by dhollman** [comment]
Basically, the assignment operator needs to "turn back on" reference counting for both the source and the destination.  We'd have to keep some sort of "residual count" in the `SharedAllocationHeader` (or some other place accessible on the device) that needs to get combined back in upon destruction (which will be tricky because the destructor of the user's functor gets called on the host and not on the device currently).

**Comment by mhoemmen** [comment]
Would this make creating subviews (without explicitly using unmanaged Views) super expensive on the GPU?  I'm curious if we would need to start making all the Views in kernels unmanaged.

**Comment by dhollman** [comment]
I’ve always thought that unmanaged views are the way to go if you care about performance. But that’s definitely something to consider. I’m not proposing we regress the performance of apps already doing this; I just want it to work and not be a landmine waiting for unfortunate foot guns.

**Comment by mhoemmen** [comment]
> and not be a landmine waiting for unfortunate foot guns.

This metaphor stretches my imagination a bit ;-P

**Comment by mhoemmen** [comment]
@dhollman I agree, though, that it would be better for apps to think harder about data ownership.

**Comment by crtrott** [comment]
Wouldn't this use the move constructor if we do it right:
```c++
auto view0_sub = Kokkos::subview (view0, std::pair {8, 11});
```

**Comment by mhoemmen** [comment]
@crtrott Do you mean for `view0_sub`?  Return value optimization may (pre C++17; must, post C++17) elide its construction, so the move constructor might (must, post C++17) never get invoked.

**Comment by dhollman** [comment]
@crtrott Keep in mind that any time you have a variable declaration on the left-hand side, you're never using an assignment operator (I forget this from time to time also).  @mhoemmen is right that RVO may cause the move constructor to never be invoked (but it has to be present anyway). This should still be irrelevant for reference counting, since `view0_sub` will get destroyed before leaving the scope of the parallel dispatch.  If `view0_sub` participates in an assignment as a right-hand side, though, it would need to "reach in" to `view0` and turn on reference counting.  This shouldn't be all that bad, since it will still have access to the same shared allocation record.

**Comment by dalg24** [comment]
Do not close.  It needs another champion.
